src/utils/manager.py

```diff
@@ -223,12 +268,23 @@ def umount(mountpoint: Union[str, os.PathLike], force: bool = False) -> None:
     Raises:
         Error: Raised if NFS share umount operation fails.
     """
-    cmd = ["umount"]
+    cmd = ["service", "autofs"]
     if force:
-        cmd.append("--force")
-    cmd.append(mountpoint)
+        cmd.append("restart")
+    else:
+        cmd.append("reload")
+
     try:
         _logger.debug(f"Unmounting NFS share at mountpoint {mountpoint}")
+
+        autofs_id = _mountpoint_to_autofs_id(mountpoint)
+        for file in [f"/etc/auto.master.d/{autofs_id}.autofs", f"/etc/auto.{autofs_id}"]:
+            try:
+                os.remove(file)
+            except OSError as e:
+                _logger.debug(f"Failed to remove file `{file}`. Reason:\n{e}")
+                raise Error(f"Failed to remove file `{file}`")
```

Comment:
We don't need to use a for loop here. Typically I would only encourage that we use one if we had a variable amount of files that we needed to perform an operation on here, but it looks like we have only two files here that we need to remove. Since that's the case, we could like simplify our implementation here greatly:

```suggestion
        pathlib.Path(f"/etc/auto.master.d/{autofs_id}.autofs").unlink(missing_ok=True)
        pathlib.Path(f"/etc/auto.{autofs_id}").unlink(missing_ok=True)
```

Since we're now using systemd instead of the mount command, we can just switch over to just handling a `SystemdError` rather than a `subprocess.TimeoutExpired` or a `subprocess.CalledProcessError`.

Reply:
I think it's fine to replace the loop with an utility function. However, I think we should still catch other OS errors so that the unit can block if deleting a file fails.

Reply:
It's fine to not catch or handle the exception here. [`OSError`](https://docs.python.org/3/library/exceptions.html#OSError) is a generic exception for system-related errors. This try-except block would catch an `OSError` here because `FileNotFoundError` inherits from `OSError`. `FileNotFoundError` is what would be raised by `.unlink()` if the file did not exist. If you set `missing_ok=False`, a `FileNotFoundError` would be raised if the autofs file did not exist. 

Is this the behavior we really want though? What we're actually concerned with is ensuring that the NFS share is no longer mounted on the host. If the autofs file doesn't exist, then restarting the autofs service should mean that the NFS share will no longer mounted on the client which is the behavior that we want. We should only block the charm if we fail to unmount the NFS share, not because we failed to delete a file. Also, there may be cases where an administrator manually deletes the autofs file. Admins can ssh into machines using Juju to to perform tasks not covered by the deployed charms. Our "happy path" is unmounting the NFS share, so even if the autofs file is not found, we're still on the happy path for unmounting the NFS share through the charm.

Let me know if you have any questions.

Reply:
I'm not worried about the files not existing, I completely agree that skipping deletion is completely fine in that case. What I'm mostly worried about is failing to deleting a file if the permissions of the file changed between mount and umount calls, for example.

Reply:
I think it's fine to continue forward with directly using `.unlink(missing_ok=True)`. I'd like to see a use case where an admin would be required to change the permissions of `etc/autofs` before we optimize for that level of error handling. The autofs files will also be owned by root since charms are executed as root, so someone with root level access would be the only one capable of changing the permissions 

If admins are randomly changing permissions on the Juju-managed machines, we have bigger problems :laughing: